### PR TITLE
feat(node): v15.9

### PR DESCRIPTION
**Pull request by SimonSchick**
Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] [Add or edit tests](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#my-package-teststs) to reflect the change.
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] [Run `npm test <package to test>`](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#running-tests).

Select one of these and delete the others:

If changing an existing definition:
- [x] Provide a URL to documentation or source code which provides context for the suggested changes: https://github.com/nodejs/node/blob/master/doc/changelogs/CHANGELOG_V15.md#2021-02-17-version-1590-current-danielleadams
- [x] If this PR brings the type definitions up to date with a new version of the JS library, update the version number in the header.

Depends on https://github.com/DefinitelyTyped/DefinitelyTyped/pull/52732

### CODE REVIEW

**Comment by SimonSchick** [comment]
Accidentally mushed together 15.8 and 15.9 commits while rebasing, hope that's not an issue.

**Comment by SimonSchick** [comment]
@borisyankov I am unsure why the github test runner fails, the errors don't really make sense to me since it is complaining about files in the DT root node_modules directory?

**Comment by SimonSchick** [comment]
@peterblazejewicz can I get some eyes one this?

**Comment by SimonSchick** [comment]
Also @sandersn as recommended by someone else.

**Comment by peterblazejewicz** [comment]
It has no, obviously, relation to PR changes:
> Error in express-serve-static-core
Error: /home/runner/work/DefinitelyTyped/DefinitelyTyped/node_modules/buffer/index.d.ts:1:1
ERROR: 1:1  no-outside-dependencies  File /home/runner/work/DefinitelyTyped/DefinitelyTyped/node_modules/buffer/index.d.ts 
TBI

**Comment by peterblazejewicz** [comment]
#53432

File: types/express-serve-static-core/index.d.ts [hunk 1]
```diff
@@ -961,7 +961,7 @@ export interface Response<
      * After middleware.init executed, Response will contain req property
      * See: express/lib/middleware/init.js
      */
-    req?: Request;
+    req: Request;
```
Comment (dnalborczyk) [inline types/express-serve-static-core/index.d.ts:964]: this is a change for `express-serve-static-core`. did you want this included in this PR?

Comment (SimonSchick) [reply types/express-serve-static-core/index.d.ts:964]: Yes, otherwise the express response type incorrectly extends the request type.
I looked into the express code and it seems practically impossible to optain a response without `req` set.

**Comment by peterblazejewicz** [comment]
@SimonSchick Iv'e just found single case, there are others. I'm able to rereate issue locally with just one single package:
```
dtslint types "safer-buffer"
```
TBU

**Comment by SimonSchick** [comment]
Oh hey, it works~

**Comment by SimonSchick** [comment]
@peterblazejewicz current workaround is to use `node:buffer` (which is supported anyways), so I think this is ready to merge!

File: types/node/readline.d.ts [hunk 2]
```diff
@@ -132,10 +140,23 @@ declare module 'readline' {
...
         output?: NodeJS.WritableStream;
         completer?: Completer | AsyncCompleter;
+        /**
+         *  Initial list of history lines. This option makes sense
+         * only if `terminal` is set to `true` by the user or by an internal `output`
+         * check, otherwise the history caching mechanism is not initialized at all.
+         * @default []
+         */
         terminal?: boolean;
+        history?: string[];
```
Comment (peterblazejewicz) [inline types/node/readline.d.ts:150]:
```suggestion
        history?: string[];
        terminal?: boolean;
```

**Review by peterblazejewicz — APPROVED** [review]

**Review by peterblazejewicz — APPROVED** [review]

**Comment by SimonSchick** [comment]
Ready to merge

**Comment by mrjackwills** [comment]
When using this updated  @types/node@15.9.0, my express app encounters overload errors with the following code;

```api.use(express.json());```

```api.all(morgan('dev'));```

where api is ```export const api = express();```

error message:
```
The last overload is declared here.
No overload matches this call.
  The last overload gave the following error.
    Argument of type 'NextHandleFunction' is not assignable to parameter of type 'PathParams'.
      Type 'NextHandleFunction' is not assignable to type '(string | RegExp)[]'.
```

File: types/node/http.d.ts [hunk 3]
```diff
@@ -156,7 +156,8 @@ declare module 'http' {
 
     // https://github.com/nodejs/node/blob/master/lib/_http_outgoing.js
     class OutgoingMessage extends stream.Writable {
-        upgrading: boolean;
+        readonly req: IncomingMessage;
+
```
Comment (Meyclem) [inline types/node/http.d.ts:160]: May I ask why this change has been introduced? I see above that you added the same kind of thing here: https://github.com/DefinitelyTyped/DefinitelyTyped/pull/52775/files#r640537682

I ask the question because I'm working on a web framework with my team and this is a breaking change of us.

**Comment by SimonSchick** [comment]
@Meyclem I saw a comment from you in my emails but cannot find it here, did you delete it? Is it still something that needs to be addressed?

**Comment by electerious** [comment]
I've updated my dependencies and now I'm getting the following error. It somehow seems to be related to this PR, but I'm not sure in which way as I'm not directly using the types (it's a dependency of a dependency. No fun to debug at all :D).

```
./../node_modules/.pnpm/@types+express-serve-static-core@4.17.19/node_modules/@types/express-serve-static-core/index.d.ts:589:18 - error TS2430: Interface 'Response<ResBody, Locals, StatusCode>' incorrectly extends interface 'ServerResponse'.
  Property 'req' is optional in type 'Response<ResBody, Locals, StatusCode>' but required in type 'ServerResponse'.

589 export interface Response<
                     ~~~~~~~~

../../node_modules/.pnpm/@types+express@4.17.11/node_modules/@types/express/index.d.ts:58:55 - error TS2344: Type 'Response<any, Record<string, any>>' does not satisfy the constraint 'ServerResponse'.
  Property 'req' is optional in type 'Response<any, Record<string, any>>' but required in type 'ServerResponse'.

58     var static: serveStatic.RequestHandlerConstructor<Response>;
```
